### PR TITLE
Refactor taint analyzer

### PR DESCRIPTION
**Pull request by ayomawdb**
## Purpose
Refactoring relevant to taint analyzer and remove irrelevant code segments no longer relevant to taint checking operations.

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java [hunk 1]
```diff
@@ -1764,51 +1720,48 @@ private void addTaintError(DiagnosticPos diagnosticPos, String paramName, Diagno
...
         TaintRecord allParamsUntaintedRecord = taintTable.get(ALL_UNTAINTED_TABLE_ENTRY_INDEX);
         if (allParamsUntaintedRecord.taintError != null && allParamsUntaintedRecord.taintError.size() > 0) {
             // This can occur when there is a error regardless of tainted status of parameters.
             // Example: Tainted value returned by function is passed to another functions's sensitive parameter.
             addTaintError(allParamsUntaintedRecord.taintError);
         } else {
-            returnTaintedStatus = new ArrayList<>(
-                    taintTable.get(ALL_UNTAINTED_TABLE_ENTRY_INDEX).retParamTaintedStatus);
+            if (taintTable.get(ALL_UNTAINTED_TABLE_ENTRY_INDEX).retParamTaintedStatus.size() > 0) {
+                returnTaintedStatus = new Boolean(taintTable.get(ALL_UNTAINTED_TABLE_ENTRY_INDEX).retParamTaintedStatus
```
Comment (anupama-pathirage) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java]: We don't need to create new Boolean here

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java [hunk 2]
```diff
@@ -1826,19 +1779,17 @@ private void analyzeInvocation(BLangInvocation invocationExpr) {
                             break;
                         }
                     }
-                    analyzeArgument(paramIndex, invokableSymbol, invocationExpr, argExpr, returnTaintedStatus);
-                    if (stopAnalysis) {
-                        break;
+                    if (analyzeArgument(paramIndex, invokableSymbol, invocationExpr, argExpr)) {
```
Comment (anupama-pathirage) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java]: defaultableParamsInitialIndex which is used to calculate paramIndex is redundant

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java [hunk 3]
```diff
@@ -218,6 +218,7 @@
     private static final String ANNOTATION_SENSITIVE = "sensitive";
 
     private static final int ALL_UNTAINTED_TABLE_ENTRY_INDEX = -1;
+    private static final int RETURN_TAINTED_STATUS_COLUMN_INDEX = 0;
 
     private enum AnalyzerPhase { INITIAL_ANALYSIS, BLOCKED_NODE_ANALYSIS };
```
Comment (anupama-pathirage) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java]: Do we need ";" at the end?

Comment (ayomawdb) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java]: Not necessary. Will remove.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java [hunk 4]
```diff
@@ -257,7 +256,15 @@ public void visit(BLangPackage pkgNode) {
         this.currPkgEnv = pkgEnv;
         this.env = pkgEnv;
 
-        pkgNode.topLevelNodes.forEach(e -> ((BLangNode) e).accept(this));
+        for (TopLevelNode e : pkgNode.topLevelNodes) {
+            try {
+                ((BLangNode) e).accept(this);
+            } catch (InvokableAnalysisBlockedOnDependencyException ignore) {
```
Comment (hasithaa) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java]: Handling Java exceptions is an expensive operation. That will slow down the analysis process. Hence shall we revert exception based logic, and improve the old status based implementation.

**Comment by ayomawdb** [comment]
@hasithaa Agreed on the extensiveness of exception based approach. I have made the suggested change with last two commits.

**Pull request merged**